### PR TITLE
fix(otel): Fix Dynamic Sampling context propagation for nested spans

### LINKED ISSUE #553: [OTel] Dynamic Sampling Context propagation does not work for nested spans
Extracted from #537.

### Problem

Dynamic Sampling Context propagation from nested spans doesn't work properly in OpenTelemetry integration.
Main reason: no publicly exposed mechanism to get the transaction (as in, a root span) for the given nested span. 
At the moment we're [doing this](https://github.com/getsentry/sentry-go/blob/2ece4821c5ab58cb10a385c8bbdc1270a5a0b432/otel/propagator.go#L45-L49), meaning that if we're doing a cross-service request from a nested span, the DSC won't be properly set.

### Potential Solutions

* Expose a method on Span that allows to get the root span (transaction) of the given span.

### PR DESCRIPTION
**Pull request by tonyo**
Thanks to #558, we can now fix the DSC/baggage propagation for non-transaction spans in the OTel propagator.

Fixes #553.

### CODE REVIEW

File: otel/propagator.go [hunk 1]
```diff
@@ -45,10 +45,12 @@ func (p sentryPropagator) Inject(ctx context.Context, carrier propagation.TextMa
...
-	if sentrySpan != nil && sentrySpan.IsTransaction() {
-		// TODO(anton): Normally, this should return the DSC baggage from the transaction
-		// (not necessarily the current span). So this might break things in some cases.
-		sentryBaggageStr = sentrySpan.ToBaggage()
+	if sentrySpan != nil {
+		// Dynamic sampling context is set only on the (root) transaction.
+		sentryTransaction := sentrySpan.GetTransaction()
+		if sentryTransaction != nil {
+			sentryBaggageStr = sentryTransaction.ToBaggage()
+		}
```
Comment (tonyo) [inline otel/propagator.go:53]: This is the actual fix.

Comment (cleptric) [reply otel/propagator.go:53]: Tbh, I would actually fix `ToBaggage()`.

```go

// ToBaggage returns the serialized dynamic sampling context in the baggage format.
func (s *Span) ToBaggage() string {
	if s.IsTransaction() {
		return s.GetTransaction().dynamicSamplingContext.String()
	}

	return s.dynamicSamplingContext.String()
}
```

File: tracing.go [hunk 2]
```diff
@@ -288,7 +288,10 @@ func (s *Span) ToSentryTrace() string {
 
 // ToBaggage returns the serialized dynamic sampling context in the baggage format.
 func (s *Span) ToBaggage() string {
-	return s.dynamicSamplingContext.String()
+	if containingTransaction := s.GetTransaction(); containingTransaction != nil {
+		return containingTransaction.dynamicSamplingContext.String()
+	}
+	return ""
```
Comment (tonyo) [inline tracing.go:294]: Moved the updated logic here.

**Comment by tonyo** [comment]
Alright, found another issue: in the propagator's `Inject` we should actually generate/freeze the baggage.
It's getting more obvious that we should refactor/redo the DSC generation and freezing logic. I think we need a more explicit `FinalizeBaggage()` (or similar) method on the transaction object. 
I don't want to do it in this PR which already got messier than I would like.

**Comment by cleptric** [comment]
What issue, in particular, you're seeing? Freezing the baggage during creation is possible today, what would a more explicit way improve?

File: helpers_test.go [hunk 3]
```diff
@@ -50,3 +52,20 @@ func formatUnequalValues(got, want interface{}) string {
...
+	baggageGot, err := baggage.Parse(got)
+	if err != nil {
+		t.Error(err)
+	}
+	baggageWant, err := baggage.Parse(want)
+	if err != nil {
+		t.Error(err)
+	}
+
+	if !reflect.DeepEqual(baggageGot, baggageWant) {
```
Comment (cleptric) [inline helpers_test.go:68]: I would always use https://pkg.go.dev/github.com/google/go-cmp/cmp, had quite a few issues with reflect in the past.

File: otel/propagator_test.go [hunk 4]
```diff
@@ -28,37 +28,31 @@ type transactionTestContext struct {
...
-		Release:       "1.2.3",
-		EnableTracing: true,
-	})
-	hub := sentry.NewHub(client, sentry.NewScope())
-	ctx := sentry.SetHubOnContext(context.Background(), hub)
-
+func createTransactionAndMaybeSpan(transactionContext transactionTestContext, withSpan bool) trace.SpanContextConfig {
 	transaction := sentry.StartTransaction(
-		ctx,
+		emptyContextWithSentry(),
```
Comment (cleptric) [inline otel/propagator_test.go:33]:
```suggestion
		emptyCtxWithSentry(),
```

File: otel/propagator_test.go [hunk 5]
```diff
@@ -28,37 +28,31 @@ type transactionTestContext struct {
...
 	transaction.TraceID = TraceIDFromHex(transactionContext.traceID)
 	transaction.SpanID = SpanIDFromHex(transactionContext.spanID)
-	transaction.SetDynamicSamplingContext(sentry.DynamicSamplingContextFromTransaction(transaction))
 
-	sentrySpanMap.Set(trace.SpanID(transaction.SpanID), transaction)
 	if withSpan {
 		span := transaction.StartChild("op")
+		// We want the child to have the SpanID from transactionContext, so
+		// we "swap" span IDs from the transaction and the child span.
+		transaction.SpanID = span.SpanID
```
Comment (cleptric) [inline otel/propagator_test.go:44]: Not sure I fully understand this.

**Review by cleptric — APPROVED** [review]
Minor asks/questions, else LGTM!

**Comment by tonyo** [comment]
I see the following issues:
1. Given a transaction, there's no easy way now to generate the DSC, and then freeze it. We need to do stuff like 
    `transaction.SetDynamicSamplingContext(DynamicSamplingContextFromTransaction(transaction))`, which is cumbersome and error-prone
2. `Span.ToBaggage()` returns the serialized DSC, but doesn't generate it. Honestly, I don't see a good case of using it alone.

> Freezing the baggage during creation is possible today, what would a more explicit way improve?

Ideally we should have convenient ways of explicitly generating the DSC/baggage in a proper format (as baggage values, or as a string). I don't have a super clear picture yet how it will look like, though.

File: helpers_test.go (hunk 3 above)
Comment (tonyo) [reply helpers_test.go:68]: Agreed, will include things like this in the planned test utils refactoring.

File: otel/propagator_test.go (hunk 5 above)
Comment (tonyo) [reply otel/propagator_test.go:44]: When `withSpan` is passed, we want to create a transation and a child span. We also want the child span to have the predefined SpanID (passed via `transactionContext`). It means that we have to update the SpanID for the transaction. We can generate a new random one, or we can just take the already-generated one from the child span (`span`), that will be replaced with `SpanIDFromHex(transactionContext.spanID)` in the next line anyway.

**Comment by cleptric** [comment]
The DSC APIs are something I want to avoid to expose to publicly, as it has lots of nitty-gritty details. Taking some extra steps for Otel is fine IMO.

> Span.ToBaggage() returns the serialized DSC, but doesn't generate it. Honestly, I don't see a good case of using it alone.

You would use this function to get the value of the header string for outgoing HTTP requests.
Here is a real-world example where I'm using it: https://github.com/getsentry/gib-potato/blob/e7eda439ad93517a27f7c3201d838b769c766587/potal/httpclient.go#L37

File: otel/propagator_test.go (hunk 4 above)
Comment (tonyo) [reply otel/propagator_test.go:33]: It's a helper function that's used in multiple places already, do we really want this?

Comment (cleptric) [reply otel/propagator_test.go:33]: Up to you, no strong feelings.

**Comment by cleptric** [comment]
You are right, we actually need to create a DSC when calling `ToBaggage` if there is none. This is the case if we're the head SDK. Great catch!

**Comment by tonyo** [comment]
Cool, will think about how to improve this (as a separate task)

**Pull request merged**